Posts: handle deleted posts from a post-list request

rralian opened the pull request:
In certain cases we get data for post-lists that overlaps with data already in memory. Because this data is fresher than the in-memory data, we may have learned that a post currently in the list should no longer be in the list. This commit handles this case and removes the post from the list.

## Testing
For now the most important thing is to make sure post-list is working correctly. Load up a page of drafts or published posts, and scroll down enough to trigger a couple of extra fetches from the api. Ensure you get the same results as you do in master on wpcalypso.

This will come in handy if/when we try to respond optimistically to post-list requests with locally cached data. When we refresh the data with a callback from the server and run the callback again, this should clear out any items that should have been deleted.

/cc @retrofox

retrofox commented on client/lib/posts/post-list-store.js:
```diff
@@ -204,8 +212,31 @@ export default function( id ) {
+	 *
+	 * @param  {array} currentList - stored list of PostIDs
+	 * @param  {array} newPosts    - new page of postIDs
+	 * @return {array}             - postIDs in currentList that should be deleted
```
I think you should tweak @return here http://usejsdoc.org/tags-returns.html

rralian replied:
I'm using docblockr (in atom) to set these, and it seems opinionated about syntax. Which aspect of @return are you taking issue with?

retrofox commented on client/lib/posts/post-list-store.js:
```diff
@@ -103,14 +103,16 @@ export default function( id ) {
-	function sort() {
+	 * sort post-list according to current orderBy parameter
+	 * @param  {array} list - list of posts to sort
+	 * @return {array}      sorted array (original will not be mutated)
```
remove spaces/tabs ? http://usejsdoc.org/tags-returns.html

rralian replied:
Can we be a little more flexible about jsdoc documentation? I'm not sure everyone has converged on http://usejsdoc.org as a standard, and I'd rather retain the use of the tooling (using docblockr) than have to ditch the tool and write the blocks manually to match the specs. I see within calypso other people seem to be using different tooling as well, as I see a number of different flavors of jsdoc, but they all seem understandable.

gwwar commented on client/lib/posts/post-list-store.js:
```diff
@@ -103,14 +103,16 @@ export default function( id ) {
+	 * @return {array}      sorted array (original will not be mutated)
+	 */
+	function sort( list ) {
 		var key = _activeList.query.orderBy;
```
const?

rralian replied:
sure

gwwar commented on client/lib/posts/post-list-store.js:
```diff
@@ -103,14 +103,16 @@ export default function( id ) {
-			var postA = PostsStore.get( a ),
-				postB = PostsStore.get( b ),
+		return list.slice( 0 ).sort( function( a, b ) {
+			var postA = PostsStore.get( a ) || a,
```
In the `|| a` case, does the post run through the normalizer?

rralian replied:
No, it's the raw Post object. I added the `|| a` because I wanted to sort the raw Post list from the server before I compare it to the stored list. But actually that's unnecessary as it arrives sorted, so I can remove that.

gwwar commented on client/lib/posts/post-list-store.js:
```diff
@@ -157,14 +159,20 @@ export default function( id ) {
 		} );
 
+		// were some posts missing from the response?
+		const removedPosts = getRemovedPosts( _activeList.postIds, postIds );
```
I think this needs to be moved down to the `if( postIds.length )` block to handle an empty array

gwwar commented on client/lib/posts/post-list-store.js:
```diff
@@ -204,8 +212,31 @@ export default function( id ) {
+	 * @param  {array} newPosts    - new page of postIDs
+	 * @return {array}             - postIDs in currentList that should be deleted
+	 */
+	function getRemovedPosts( currentList, newPosts ) {
```
It would be helpful to add unit tests for this function, since this is a strange case.

gwwar replied:
Also add a quick return if currentList is < 2.

gwwar commented on client/lib/posts/post-list-store.js:
```diff
@@ -204,8 +212,31 @@ export default function( id ) {
+	 */
+	function getRemovedPosts( currentList, newPosts ) {
+		const overlapBegin = currentList.indexOf( newPosts[0] );
+		const overlapEnd = currentList.indexOf( newPosts.slice( -1 ) );
```
This currently returns an array. You probably mean `newPosts.slice( -1 )[ 0 ]`

gwwar commented:
At a minimum, you'll need to update the newPosts.slice( -1 ) line, so that this function returns anything other than an empty array. I would really suggest moving getRemovedPosts to an area where you can unit test it. After you get that change in, it should be ok to :ship:

rralian commented:
> I would really suggest moving getRemovedPosts to an area where you can unit test it. After you get that change in, it should be ok to :ship:

You're right, was just thinking I should pull this down and get it unit-tested.

rralian commented:
Thanks for the reviews @retrofox / @gwwar. Sorry, that was sloppy/lazy of me not to provide tests. I've added a unit test and a lifecycle test. I've also refactored with your suggestions. I'm hoping @retrofox will forgive me the loose jsdoc syntax. :smile:

retrofox commented:
x-(

gwwar commented:
:shipit:
